specification.md

```diff
@@ -2201,6 +2219,169 @@ authenticity of the downloaded file, for example can be the
+
+### CustomMessage
+
+CustomMessage allows for custom messages to be sent between the Agent and Server. It is supported on ServerToAgent and
+AgentToServer. It requires that the Agent and Server both agree on the contents and encoding of the message. If the type
+is not recognized, the message MUST be ignored and a reply with CustomMessageAck with Status NOT_SUPPORTED MUST be sent.
+
+A CustomMessage identifies the type of message and can contain a string body, binary data, neither, or both. It also contains a
+sequence_num that is used to associated a CustomMessage with a [CustomMessageAck](#CustomMessageAck).
+
+```protobuf
+message CustomMessage {
+    uint64 sequence_num = 1;
+    string type = 2;
+    string body = 3;
+    bytes data = 4;
+}
+```
+
+#### CustomMessage.sequence_num
+
+The sequence_num of this CustomMessage so that the CustomMessageAck can be associated with this message. This
+number MUST be increased with each message to allow messages to be uniquely identified. In the case of a
+CustomMessage in an AgentToServer message, this MAY be the sequence_num of the enclosing AgentToServer message.
+
+#### CustomMessage.type
+
+The message type should a reverse FQDN that uniquely identifies the custom message type.
+
+#### CustomMessage.body
+
+Optional string body of the message. The Agent and Server must agree on the format of the contents.
+
+#### CustomMessage.data
+
+Optional binary data of the message. The Agent and Server must agree on the format of the contents.
+
+### CustomMessageAck
+
+CustomMessageAck MUST be returned in response to a CustomMessage. It allows the sender of the CustomMessage to determine
+if the message was successfully handled.
+
+Handling a CustomMessage may result in another CustomMessage in response, e.g. a Server could request information from
+an Agent and expect the Agent to reply with that information in a separate CustomMessage. The CustomMessageAck allows
+the sender to know if it should expect a reply or if the message was ignored or was an error.
+
+```protobuf
+message CustomMessageAck {
+    enum Status {
+        NOT_SUPPORTED = 0;
+        OK = 1;
+        ERROR = 2;
+    }
+    uint64 msg_sequence_num = 1;
+    Status status = 2;
+    string error_message = 3;
+}
+```
+
+#### CustomMessageAck.msg_sequence_num
+
+This Ack is in response to the CustomMessage with the corresponding sequence_num.
+
+#### CustomMessageAck.status
+
+The status field indicates if and how the custom message was handled. Possible values are:
+
+NOT_SUPPORTED: The CustomMessage was ignored because the message type was not recognized or supported. The error_message may contain additional information.
+
+OK: The CustomMessage was handled successfully.
+
+ERROR: The CustomMessage was unable to be handled successfully. The error_message should contain additional information.
+
+#### CustomMessageAck.error_message
+
+Error message in string form, typically human readable.
```

Comment:
Seems like this format is arbitrary and left up to agreement based on the specific Agent/Server? Only asking because we call out `typically human readable` which might not be necessary if the receiver of the Ack knows how to parse it and can present it to the user in a useful way.